_overviews/scala3-contribution/arch-time.md

```diff
@@ -0,0 +1,100 @@
+---
+title: Time in the Compiler
+type: section
+description: This page describes the concepts of time in the Scala 3 compiler.
+num: 16
+previous-page: arch-types
+next-page: arch-symbols
+---
+
+> (The following is work in progress), adapted from dotty.epfl.ch
+
+Conceptually, the `dotc` compiler's job is to maintain views of various
+artifacts associated with source code at all points in time.  But what is
+*time* for `dotc`? In fact, it is a combination of compiler runs and compiler
+phases.
+
+The *hours* of the compiler's clocks are measured in compiler [runs]. Every run
+creates a new hour, which follows all the compiler runs (hours) that happened
+before. `dotc` is designed to be used as an incremental compiler that can
+support incremental builds, as well as interactions in an IDE and a REPL. This
+means that new runs can occur quite frequently.  At the extreme, every
+keystroke in an editor or REPL can potentially launch a new compiler run, so
+potentially an "hour" of compiler time might take only a fraction of a second
+in real time.
+
+The *minutes* of the compiler's clocks are measured in phases. At every
+compiler run, the compiler cycles through a number of [phases]. The list of
+phases is defined in the [Compiler] object There are currently about 60 phases
+per run, so the minutes/hours analogy works out roughly. After every phase the
+view the compiler has of the world changes: trees are transformed,  types are
+gradually simplified from Scala types to JVM types, definitions are rearranged,
+and so on.
+
+Many pieces in the information compiler are time-dependent. For instance, a
+Scala symbol representing a definition has a type, but that type will usually
+change as one goes from the higher-level Scala view of things to the
+lower-level JVM view. There are different ways to deal with this. Many
+compilers change the type of a symbol destructively according to the "current
+phase". Another, more functional approach might be to have different symbols
+representing the same definition at different phases, which each symbol
+carrying a different immutable type. `dotc` employs yet another scheme, which
+is inspired by functional reactive programming (FRP): Symbols carry not a
+single type, but a function from compiler phase to type. So the type of a
+symbol is a time-indexed function, where time ranges over compiler phases.
+
+Typically, the definition of a symbol or other quantity remains stable for a
+number of phases. This leads us to the concept of a [period]. Conceptually,
```

Comment:
I'm surprised to see so much detail about `Period`, which I've never used/manipulated in practice. Conversely, I think it would be extremely valuable to have information about *time travel* with `atPhase`, which is used in many places in the codebase, and is currently missing.

Reply:
I benefitted from this info, seeing as it's new in dotc.  But I don't disagree that phase travelling is even more important and I've never seen a good description of it.

Reply:
This was copy-pasted from the original dotty website, but this would be good to add